### ISSUE TITLE
Prevent brute force attack on admin login

### ISSUE THREAD

**Issue by hadifarnoud** [strapi/strapi#9470]
Strapi version: 3.4.6

It's possible to do brute force attack on Strapi admin login. Currently, there are no way of rate limiting in Strapi for login.

**Comment by hadifarnoud** [comment]
types of attacks possible:

CWE-307: Improper Restriction of Excessive Authentication Attempts
CAPEC-112: Brute Force

CVSS 7.5

path: /documentation/login
path: /admin/auth/login

how to fix this issue? captcha should show up after a few failed login attempts

**Comment by derrickmehaffy** [comment]
We could do something simple like we do with Users-Permissions: https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/config/policies/rateLimit.js

**Comment by alexandrebodin** [comment]
Hi @hadifarnoud Can you give more details on the security level you would expect ? a backoff system ? based on ip ? fail2ban ?
We can add a simple layer in memory but doesn't feel like a really world solution to me. 
I would always recommend doing this on the proxy or api gatway layer. e.g https://gist.github.com/joecampo/848178ab5c18aada0eab nginx with fail2ban

**Comment by hadifarnoud** [comment]
@alexandrebodin nginx with fail2ban would not fix this issue as attacker can change IP and try again. it should be done on Strapi level and show captcha or block access on user level not IP

**Comment by alexandrebodin** [comment]
So you would restrict based on email ? that wouldn't really stop a bruteforce attack on multi users.

**Comment by hadifarnoud** [comment]
@alexandrebodin if a captcha shows up on third attempt and block requests without solved captcha, that would solve the issue completely. attacker cannot guess the account password because the captcha limitation is activated.

the way bruteforce attack works it needs thousands of request to eventually find the right password.

**Comment by hadifarnoud** [comment]
that's basically how Yahoo deals with the bruteforce attack for example. you can test it yourself

**Comment by alexandrebodin** [comment]
Thanks for the details @hadifarnoud, 

Your proposed solution will only work if we start persisting into db the login attempts. I was trying to see if your suggestion could work with a backoff strategy to avoid adding too much work & UI changes.

I'll share this with the PMs

**Comment by derrickmehaffy** [comment]
Yeah we would need to persist sessions in the DB for a solution like that. Most captcha services also require a key (Google's is no longer free) also.

**Comment by hadifarnoud** [comment]
@derrickmehaffy reCaptcha is "Free up to 1 million Assessments / Month*" which is plenty for many users.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy reCaptcha is "Free up to 1 million Assessments / Month*" which is plenty for many users.

But it requires a connection to an outside resource which generally goes against our base rules.

Given we could make it optional but if users don't opt-in then it doesn't really solve the issue.

And if they change those limits or remove the free option entirely we would have to switch.

**Comment by dalbitresb12** [comment]
@derrickmehaffy Another option could be to use [hCaptcha](https://www.hcaptcha.com/) instead of reCaptcha.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy Another option could be to use [hCaptcha](https://www.hcaptcha.com/) instead of reCaptcha.

Still has a "free" and paid plan, and is still a service that every user would have "sign up" for. 

It's highly unlikely we will include any feature that by default relies on an external service. The whole point of Strapi is to give control of your data to the user who runs it. By using an external service, that user no longer has control and in this case its security.

**Comment by stafyniaksacha** [comment]
Another way to improve security is to implement a sleep method on login API request. 
The sleep time can be increased by a power of number of failed attempts and reset on success login.

We can lock login access to this user and deliver an email with a token (like for registration) to unlock this access

**Comment by derrickmehaffy** [comment]
@stafyniaksacha yeah it's the same as a rate-limit (though in a different style like a leaky bucket method) with the extra steps of a lock and notify.

The same issue exists here as it would be with just a normal rate limit. We need to store that rate-limit information in the database and not node's memory.

**Comment by hadifarnoud** [comment]
> > @derrickmehaffy Another option could be to use [hCaptcha](https://www.hcaptcha.com/) instead of reCaptcha.
> 
> Still has a "free" and paid plan, and is still a service that every user would have "sign up" for.
> 
> It's highly unlikely we will include any feature that by default relies on an external service. The whole point of Strapi is to give control of your data to the user who runs it. By using an external service, that user no longer has control and in this case its security.

it's better than being in danger of getting hacked and lose all their data. It could be a plugin and if user is using Strapi with public access, they most likely need to enable it.

I don't know any other option to be honest. Only way to fix this issue is captcha, no?

**Comment by dalbitresb12** [comment]
> The same issue exists here as it would be with just a normal rate limit. We need to store that rate-limit information in the database and not node's memory.

Is the rate limiting that is used for the normal API routes saved in the database or in Node's memory?

In any case, we can use what @stafyniaksacha said. Rate-limit until the account is locked (using the `blocked` property that already exists in the admin user model?). If the account gets locked, send the email with the verification code (that would be another property in the database, maybe `unlockToken`?) and then for the user to access their account again, just open the link in their email and the account gets unblocked (this is kind of the same as the email verification process for regular users).

@derrickmehaffy About adding a Captcha (reCaptcha or hCaptcha): if this is something that would be against the base rules of Strapi, it could be added as an optional plugin, or by extending the admin panel login page (is this possible?) the same way we can extend regular Strapi extensions (like the `users-permissions` plugin, etc.). This way at least we could add hCaptcha or reCaptcha even if not officially supported by Strapi (with a plugin or natively).

**Comment by stafyniaksacha** [comment]
hCaptcha / reCaptcha does not protect the API

**Comment by derrickmehaffy** [comment]
> hCaptcha / reCaptcha does not protect the API

Exactly, a rate-limit is backend based and thus doesn't rely on a frontend requirement

> Edit: @derrickmehaffy did you know if JWT token are stored in DB ? If none it would be easy to forge one.

@stafyniaksacha they are not, but you can't forge one without the JWT secret, I suppose in theory you could brute force this but the longer and more complex your JWT secret is, the harder and longer it would take to forge.

---

> Is the rate limiting that is used for the normal API routes saved in the database or in Node's memory?

Node Memory currently

> it could be added as an optional plugin, or by extending the admin panel login page (is this possible?) the same way we can extend regular Strapi extensions (like the users-permissions plugin, etc.). This way at least we could add hCaptcha or reCaptcha even if not officially supported by Strapi (with a plugin or natively).

In theory yes, but the complexity at the moment might make that extremely difficult. Modifications to the -core- admin (`strapi-admin` package) is not as straight forward at the moment

> Edit: If a Captcha is added extending the admin panel as I said (if possible), we should also be able to customize the controllers for the admin panel login API route. Is this possible with the current version of Strapi?

No, not without forking the entire `strapi-admin` package. We have another issue about turning this package into a plugin to allow for backend customizations: https://github.com/strapi/strapi/issues/4838

> I don't know any other option to be honest. Only way to fix this issue is captcha, no?

Proper rate-limits with a blacklist/whitelist stored in the database and hooks to disable accounts that exceed a certain limit and require either a manual or automatic validation to restore their access.

---

**reCaptcha is not a viable option, it is not something we will consider due to external requirements**; the short term options would be adding some complex rate-limiting options.

**Comment by takeaction-git** [comment]
What about the possibility of using 2FA either with SMS or a service like Authy or Google Authenticator?

**Comment by hadifarnoud** [comment]
that's an excellent idea @takeaction-git. maybe just 2FA with something like Google Authenticator app

**Comment by derrickmehaffy** [comment]
https://portal.productboard.com/strapi/1-roadmap/c/43-2fa-via-users-permissions-plugin

**Comment by derrickmehaffy** [comment]
Something we could look at for the admin panel (not users-permissions right now since we aren't using passport there) is this: http://www.passportjs.org/packages/passport-2fa-totp/

The Admin panel uses passport for authentication

**Comment by derrickmehaffy** [comment]
Hey all, we are currently moving many feature requests to our new feedback and feature request website to help clean up the GitHub issues and make it easier for us to review bug reports and fix them.

I have moved this feature to the following URL: https://feedback.strapi.io/security/p/support-2fa-via-users-permissions-plugin
If you are interested in this feature please feel free to go there and add more comments and/or upvote it there.

For now I will close this and lock it so that all new information goes into our new feedback website.
Thanks!
